File: libraries/ndf/ndf_xstat.f
      SUBROUTINE NDF_XSTAT( INDF, XNAME, THERE, STATUS )
*+
*  Name:
*     NDF_XSTAT

*  Purpose:
*     Determine if a named NDF extension exists.

*  Language:
*     Starlink Fortran 77

*  Invocation:
*     CALL NDF_XSTAT( INDF, XNAME, THERE, STATUS )

*  Description:
*     The routine returns a logical value indicating whether a named
*     extension is present in an NDF.

*  Arguments:
*     INDF = INTEGER (Given)
*        NDF identifier.
*     XNAME = CHARACTER * ( * ) (Given)
*        Name of the extension.
*     THERE = LOGICAL (Returned)
*        Whether the extension is present in the NDF.
*     STATUS = INTEGER (Given and Returned)
*        The global status.

*  Algorithm:
*     -  Import the NDF identifier.
*     -  Check the extension name.
*     -  Obtain an index to the data object entry in the DCB.
*     -  Ensure that extension information is available in the DCB.
*     -  If an extension (MORE) structure does not exist, then neither
*     does the extension component.
*     -  Otherwise, search for the component to see if it is there.

*  Copyright:
*     Copyright (C) 2009 Science & Technology Facilities Council.
*     Copyright (C) 1989 Science & Engineering Research Council.
*     All Rights Reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or
*     modify it under the terms of the GNU General Public License as
*     published by the Free Software Foundation; either version 2 of
*     the License, or (at your option) any later version.
*
*     This program is distributed in the hope that it will be
*     useful,but WITHOUT ANY WARRANTY; without even the implied
*     warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
*     PURPOSE. See the GNU General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     {enter_new_authors_here}

*  History:
*     19-SEP-1989 (RFWS):
*        Original version.
*     26-SEP-1989 (RFWS):
*        Completed prologue and corrected minor errors.
*     25-NOV-2009 (TIMJ):
*        Include the name of the extension in the error message
*     {enter_further_changes_here}

*  Bugs:
*     {note_any_bugs_here}

*-

*  Type Definitions:
      IMPLICIT NONE              ! No implicit typing

*  Global Constants:
      INCLUDE 'SAE_PAR'          ! Standard SAE constants
      INCLUDE 'DAT_PAR'          ! DAT_ public constants
      INCLUDE 'NDF_PAR'          ! NDF_ private constants
      INCLUDE 'NDF_CONST'        ! NDF_ public constants

*  Global Variables:
      INCLUDE 'NDF_DCB'          ! NDF_ Data Control Block
*        DCB_XLOC( NDF__MXDCB ) = CHARACTER * ( DAT__SZLOC ) (Read)
*           Locator to extension structure.

      INCLUDE 'NDF_ACB'          ! NDF_ Access Control Block
*        ACB_IDCB( NDF__MXACB ) = INTEGER (Read)
*           Index to data object entry in the DCB.

*  Arguments Given:
      INTEGER INDF
      CHARACTER * ( * ) XNAME

*  Arguments Returned:
      LOGICAL THERE

*  Status:
      INTEGER STATUS             ! Global status

*  Local Variables:
      INTEGER IACB               ! Index to NDF entry in the ACB
      INTEGER IDCB               ! Index to data object entry in the DCB

*.

*  Check inherited global status.
      IF ( STATUS .NE. SAI__OK ) RETURN

*  Import the NDF identifier.
      CALL NDF1_IMPID( INDF, IACB, STATUS )

*  Check the extension name.
      CALL NDF1_CHXNM( XNAME, STATUS )
      IF ( STATUS .EQ. SAI__OK ) THEN

*  Obtain an index to the data object entry in the DCB.
         IDCB = ACB_IDCB ( IACB )

*  Ensure that extension information is available in the DCB.
         CALL NDF1_DX( IDCB, STATUS )
         IF ( STATUS .EQ. SAI__OK ) THEN

*  If an extension structure does not exist, then neither does the
*  extension component.
            IF ( DCB_XLOC( IDCB ) .EQ. DAT__NOLOC ) THEN
               THERE = .FALSE.

*  Otherwise, see if the specified component is present.
            ELSE
               CALL DAT_THERE( DCB_XLOC( IDCB ), XNAME, THERE, STATUS )
            END IF
         END IF
      END IF

*  If an error occurred, then report context information and call the
*  error tracing routine.
      IF ( STATUS .NE. SAI__OK ) THEN
         CALL MSG_SETC( 'NAM', XNAME )
         CALL ERR_REP( 'NDF_XSTAT_ERR',
     :   'NDF_XSTAT: Error determining if an NDF extension named ' //
     :   '"^NAM" exists.', STATUS )
         CALL NDF1_TRACE( 'NDF_XSTAT', STATUS )
      END IF

      END

File: libraries/ndf/ndfXstat.c
#include "sae_par.h"
#include "dat_par.h"
#include "ndf1.h"
#include "ndf.h"
#include "mers.h"

void ndfXstat_( int indf, const char *xname, hdsbool_t *there, int *status ){
/*
*+
*  Name:
*     ndfXstat

*  Purpose:
*     Determine if a named NDF extension exists.

*  Synopsis:
*     void ndfXstat( int indf, const char *xname, hdsbool_t *there,
*                    int *status )

*  Description:
*     This function returns a logical value indicating whether a named
*     extension is present in an NDF.

*  Parameters:
*     indf
*        NDF identifier.
*     xname
*        Pointer to a null terminated string holding the name of the
*        extension.
*     *there
*        Returned holding the whether the extension is present in the NDF.
*     *status
*        The global status.

*  Copyright:
*     Copyright (C) 2018 East Asian Observatory
*     All rights reserved.

*  Licence:
*     This program is free software; you can redistribute it and/or modify
*     it under the terms of the GNU General Public License as published by
*     the Free Software Foundation; either version 2 of the License, or (at
*     your option) any later version.
*
*     This program is distributed in the hope that it will be useful,but
*     WITHOUT ANY WARRANTY; without even the implied warranty of
*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
*     General Public License for more details.
*
*     You should have received a copy of the GNU General Public License
*     along with this program; if not, write to the Free Software
*     Foundation, Inc., 51 Franklin Street,Fifth Floor, Boston, MA
*     02110-1301, USA

*  Authors:
*     RFWS: R.F. Warren-Smith (STARLINK)
*     DSB: David S. Berry (EAO)

*  History:
*     xxx (DSB):
*        Original version, based on equivalent Fortran function by RFWS.

*-
*/

/* Local Variables: */
   NdfACB *acb;          /* Pointer to NDF entry in the ACB */
   NdfDCB *dcb;          /* Pointer to data object entry in the DCB */

/* Check inherited global status. */
   if( *status != SAI__OK ) return;

/* Ensure the NDF library has been initialised. */
   NDF_INIT( status );

/* Import the NDF identifier. */
   ndf1Impid( indf, &acb, status );

/* Check the extension name. */
   ndf1Chxnm( xname, 1, 0, status );
   if( *status == SAI__OK ) {

/* Obtain an index to the data object entry in the DCB. */
      dcb = acb->dcb;

/* Ensure that extension information is available in the DCB. */
      ndf1Dx( dcb, status );
      if( *status == SAI__OK ) {

/* If an extension structure does not exist, then neither does the
   extension component. */
         if( !dcb->xloc ) {
            *there = 0;

/* Otherwise, see if the specified component is present. */
         } else {
            datThere( dcb->xloc, xname, there, status );
         }
      }
   }

/* If an error occurred, then report context information and call the
   error tracing function. */
   if( *status != SAI__OK ) {
      msgSetc( "NAM", xname );
      errRep( " ", "ndfXstat: Error determining if an NDF extension named "
              "\"^NAM\" exists.", status );
      ndf1Trace( "ndfXstat", status );
   }

/* Restablish the original AST status pointer */
   NDF_FINAL

}

